containerfixture/containerfixture.go

```diff
@@ -0,0 +1,143 @@
+	templatedYAML := buf.String()
+	deployment.Yaml = templatedYAML
+
+	return deployment
+}
+
+// Building and Running a k8s manifest using 'podman kube play' command.
+// It will make up to three attempts to run with different configurations,
+// allowing it to run concurrently if necessary
+func BuildAndRunPod(t *testing.T, manifestTemplate string) FixtureInfo {
+	var deployment FixtureInfo
+	for attempt := 0; attempt < containerBuildAttempts; attempt++ {
+		deployment = GenerateFixturePodYaml(manifestTemplate)
+		cmd := exec.Command("podman", "play", "kube", "-")
+		cmd.Stdin = strings.NewReader(deployment.Yaml)
+		var stdoutBuf, stderrBuf bytes.Buffer
+		cmd.Stdout = &stdoutBuf
+		cmd.Stderr = &stderrBuf
+
+		err := cmd.Run()
+		if err == nil {
+			t.Logf("STDOUT: %s", stdoutBuf.String())
+			return deployment
+		}
+		t.Logf("Failed to start pod (attempt %d of %d).\n", attempt+1, containerBuildAttempts)
+		t.Logf("STDERR: %s", stderrBuf.String())
+	}
+	Cleanup(t, deployment.PodName)
+	t.Fail()
+	t.Logf("Failed to start pod after %d attempts", containerBuildAttempts)
+	return FixtureInfo{} // Avoid 'missing return' error
+}
+
+// RequireServiceIsUp continuously monitors the service container to
+// ensure it is up and ready for use in tests. It achieves this by repeatedly
+// checking the status endpoint of the service API over a period of two minutes.
+func RequireServiceIsUp(t *testing.T, endpoint string, msgAndArgs ...interface{}) {
+	netrcPath, _ := GetNetrcPath()
+	timeoutStart := time.Now().Unix()
+	for {
+		cmd := exec.Command("curl", "-s", "-o", "/dev/null", "-w", "%{http_code}",
+			"--netrc-file", netrcPath, "--insecure", endpoint)
+		output, err := cmd.Output()
+		if err != nil && string(output) != "000" {
+			t.Log(string(output))
+			t.Fatalf("curl command failed: %s", err)
+		}
+		if string(output) == "200" {
+			break
+		}
+		if time.Now().Unix()-timeoutStart > serviceUpTimeout {
+			t.Fatalf("%s", msgAndArgs)
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+// GetNetrcPath allows us to get the full path to the containerfixture
+// .netrc file from every script in the project
+func GetNetrcPath() (string, error) {
+	_, filename, _, ok := runtime.Caller(0)
+	if !ok {
+		return "", fmt.Errorf("failed to find the path of the current script")
+	}
+	dirPath := filepath.Dir(filename)
+	netrcPath := filepath.Join(dirPath, ".netrc")
+
+	return netrcPath, nil
+}
+
+// Making sure to stop and remove the pod we deployed
+func Cleanup(t *testing.T, podName string) {
+	cmd := exec.Command("podman", "pod", "stop", podName)
+	if err := cmd.Run(); err != nil {
+		t.Errorf("Error stopping the pod: %s", err)
+	}
+	removePodCmd := exec.Command("podman", "pod", "rm", "-f", podName)
+	if err := removePodCmd.Run(); err != nil {
+		t.Fatalf("Error removing the pod: %s", err)
+	}
+}
+
+func WithServiceContainer(t *testing.T, manifest string, test_func func(*testing.T, FixtureInfo), msgAndArgs ...interface{}) {
```

Comment:
the functionality of bringing a container up and down seems to be quite unrelated to the "up" check which needs to know details about the service API. The only relates detail between the two is the port number.

I would decouple them - and bring them together in the service-specific fixtures. That would allow for more flexibility in the long term such as having containers with multiple services.

Reply:
I get why you'd want to decouple the pod itself from the service (although it is specifically a service container fixture). 
The way I see for this decoupling is making `WithServiceContainer` a very simple fixture to create the pod and get the deployment data, 
while `WithSplunkContainer` run the former and does all the heavy lifting (service check, cleanup, run tests). 
Did you have anything else in mind?

Reply:
since the service details such as what the api looks like, how many port etc. is a service-specific thing - we can have the service  fixture put the checking and the bringing up/down together while keeping those as separate functions in the generic fixture module.